### ISSUE TITLE
`Realm is not defined` due to `collection-methods.js` in Jest and while Remote Debugging

### ISSUE THREAD

**Issue by liamjones** [realm/realm-js#4029]
### How frequently does the bug occur?

All the time

### Description

Upgraded from Realm 10.4.0 to 10.5.0+ (still present in 10.8.0)

Can no longer run Jest tests that use Realm or connect the Chrome Remote Debugger to the JSC, both error out with `Realm is not defined`.

This was fixed in PR #3748 but that was closed without merging. Other bug reports for this issue include issue #3835 and issue #3991.

For now, we have patched Realm via [patch-package](https://github.com/ds300/patch-package) with the code review suggestion on PR #3748 to work around this issue with the current implementation. Patch file: https://gist.github.com/liamjones/cca9c81f26e0d998c9904d734e9be400

### Stacktrace & log output

```shell
IllegalStateError: ReferenceError: Realm is not defined
    at loadFileUploadConfiguration$ (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:230110:84)
    at tryCatch (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:28789:19)
    at Generator.invoke [as _invoke] (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:28962:24)
    at Generator.next (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:28832:23)
    at next (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181472:29)
    at proc (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181447:5)
    at runForkEffect (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181688:21)
    at runEffect (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181573:770)
    at next (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181476:11)
    at currCb (http://localhost:8081/packages/apps/the-app/index.bundle?platform=ios&dev=true&minify=false:181550:9)
```

### Can you reproduce the bug?

Yes, always

### Reproduction Steps

Unfortunately, I haven't been able to make a minimal repro that still triggers the issue yet and I don't really want to give you the entire source code to our app if I can help it.

For Jest, I thought it _might_ be caused by the fact we normally have a manual mock for Realm in Jest (e.g. `__mocks__/realm.ts`) and then `jest.unmock('realm')` in one specific test to check our migration functions. However, just setting that up in a fresh RN project doesn't trigger the issue.

Setting a breakpoint in the remote debugger on https://github.com/realm/realm-js/blob/v10.8.0/lib/collection-methods.js#L21 I can see that our RN app entry point `index.js` imports our `App.tsx` and, as a result of dependencies, this eventually does `import Realm from 'realm'` and when `collection-methods.js` is imported, Realm is already not defined (but it doesn't care at this point). Later on, a redux-saga is trying to load stuff from the DB, which results in https://github.com/realm/realm-js/blob/v10.8.0/lib/collection-methods.js#L60 executing and falling over due to Realm not being defined.

### Version

10.5.0, 10.8.0

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

Yes, using encryption

### Platform OS and version(s)

macOS 11.6

### Build environment

Which debugger for React Native: React Native Debugger 0.12.1
RN 0.63.4

**Comment by liamjones** [comment]
> Would you be interested in moving from React Native Debugger to Flipper to resolve this issue?

Unfortunately, that's not possible for us right now. Flipper's debugger is only for Hermes (AFAIK) and we're stuck on the JSC at the moment. We _would_ ultimately like to try moving our apps over to Hermes and then we _have_ to use Flipper's debugger I believe?

Currently, we're blocked by Realm not supporting Hermes (I know you're working on it). Beyond that, I don't know how many other things will need changes when we try to migrate engines - we haven't been able to test yet because of the Realm blocker - and ours is a pretty chunky app so it may not be an easy thing to do if we find a lot of issues. I'm hopeful though! 🤞

> Once that is officially out, we will most likely discontinue our support of the (now legacy) Chrome Remote Debugger workflow and in that light, we probably won't be investigating this issue any further.

How will JSC based projects debug their apps containing Realm in future then? Is there another option I'm not aware of?

**Comment by liamjones** [comment]
> We are currently recommending Safari for debugging. It requires a small addition to the AppDelegate.m, but it is much more reliable than Chrome, since we don't need to maintain a workaround to communicate with the browser and the device/simulator.

@takameyer I gave this a try today and the app exits/crashes as soon as I attempt to attach Safari to debug so this doesn't look like it's a solution for us on the JSC. 😞

**Comment by fronck** [comment]
This issue should be fixed by #4109 .
Thanks for reporting.
